test/e2e/GenerateRepository.php

```diff
@@ -98,4 +105,37 @@ static function (string $path): string {
 
         return $installationTargetPath;
     }
+
+    /**
+     * @param string[] $vendorDependencies
+     */
+    private static function addVersionToDependencies(array $vendorDependencies): void
+    {
+        $composerLockPath = __DIR__ . '/../../composer.lock';
+
+        $composerLockPackages = json_decode(file_get_contents($composerLockPath))->packages;
+
+        foreach ($vendorDependencies as $dependencyPath) {
+            $composerJson = json_decode(file_get_contents($dependencyPath . '/composer.json'));
```

Comment:
Wait... Why not simply add the `composer.json` directly to those files? :thinking:

Reply:
not sure I get what you mean... which files?

Reply:
There are multiple `composer.json` files under assets, unless this modifies something else, and I'm misunderstanding your code?

Reply:
I'm probably missing something, so I'll try to recap the whole issue and my solution, so you can point to the not convincing part.

Currently the `composer.json` of the generated repositories declares as dependencies the paths to the packages already present in the `vendor` folder.
When `composer install` is executed, `composer` does a `cd` into each one of the specified paths in `vendor` and tries to find out the version of the dependency to check if it is compatible with the requirements; if the `version: ...` is missing in the `composer.json` of the dependency, then `composer` does a `git branch` to detect the available versions and doing this it gets the branches of the `you-are-using-it-wrong` package. This cause the `composer install` to fail.

The solution I implemented to this issue is to write the `version: ...` into the `composer.json` of each one of the `vendor` dependencies, so that `composer` can detect the correct version of the package.

Another option could be copying directly the whole vendor into the newly created repository, and possibly edit the `composer.json` files there, so that `composer` already finds the packages and has nothing new to install

Reply:
What I was thinking of, is to do something much simpler: right now, the code in `GenerateRepository::generateRepository()` does in https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/e2e/GenerateRepository.php#L50-L55:

```php
                    'require'           => array_merge(
                        ['roave/you-are-using-it-wrong' => 'dev-' . $currentGitVersion],
                        ...array_map(static function (string $dependency) use ($currentGitVersion): array {
                            return [$dependency => 'dev-' . $currentGitVersion];
                        }, $dependencies)
                    ),
```

According to your explanation, this is not a problem:

```php
                        ['roave/you-are-using-it-wrong' => 'dev-' . $currentGitVersion],
```

While this is:

```php
                        ...array_map(static function (string $dependency) use ($currentGitVersion): array {
                            return [$dependency => 'dev-' . $currentGitVersion];
                        }, $dependencies)
```

That's because each of the dependencies is then scanned for a "version" that we don't really have.

Now, the possible "repositories" are:

https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/e2e/GenerateRepository.php#L78-L85

These are only used in tests, so my idea was to add `"version": "1.0.0"` to each of them in their respective `composer.json` files (manually: no runtime code). That means editing:

 * https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/repositories/empty-repository/composer.json#L1-L3
 * https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/repositories/project-with-violations/composer.json#L1-L8
 * https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/repositories/repository-depending-on-type-checks/composer.json#L1-L11
 * https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/repositories/repository-indirectly-depending-on-type-checks/composer.json#L1-L11
 * https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/repositories/repository-with-dist-config/composer.json

Reply:
that doesn't work because the version issue is not relative to those repositories, but to other `vendor` libraries such as `psalm`. What we need to to is to add the version to the `psalm` `composer.json` file, so that `composer` can detect the correct version

Reply:
> What we need to to is to add the version to the `psalm` `composer.json` file, so that `composer` can detect the correct version

Those use packagist as usual? That's why this comment is in place: 

https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/e2e/GenerateRepository.php#L57-L59

Reply:
yes, Packagist is used to determine the dependencies of `you-are-using-it-wrong`. Among those, there is for example `vimeo/psalm: "^3.13.1"`. At this point `composer` finds in the declared repositories https://github.com/Roave/you-are-using-it-wrong/blob/7dc76acbce7dc231d45b8e017008fad65c7c319c/test/e2e/GenerateRepository.php#L69 that `psalm` is available there, so it does not download it but it tries to copy it from the `vendor` folder. But at this point it realises that the version of `psalm` present in the repository (computed doing a `git branch` there) does not satisfy the requirements, and so it fails. Writing the `version` in the `psalm` `composer.json` avoids this issues providing to `composer` the correct version

Reply:
> ble there, so it does not download it but it tries to copy it from the vendor folder.

This seems wrong 🤔

Reply:
might be a Composer2 bug, but that's how it seems to works at the moment

Reply:
Let's clarify with a comment for now then, and I'll merge afterwards :+1:

Reply:
I added a comment to clarify this. But now the build breaks due to a composer bug 😢

Reply:
That's some serious bad luck :S

Reply:
I opened an issue on Composer https://github.com/composer/composer/issues/9118